Simplify address matching by providing a Matcher

klingtnet opened the pull request:
A Matcher is constructed from an OSC address pattern and will use a
number of precompiled regular expression to match against a concrete OSC
address.  Regular expression match quite well to the pattern rules given
by the OSC specification and are used by other popular OSC libraries for
this purpose, e.g. [python-osc][1].

I chose `nom` as a parser combinator library to parse OSC addresses and
OSC address patterns reliably and efficiently.

@DrLuke please do not take this as a criticism of your implementation, I
am very about happy your contribution!  Actually, I would be happy if
_you_ could review this change so we can merge it back into your PR and
eventually release this feature.

My GitHub feed showed me that @D1plo1d is working on a [`no_std`
feature][2] which also adds `nom`.  I would be happy to receive a review
from @D1plo1d as well, hoping to reduce merge conflicts with their
feature.

Please be kind, I haven't programmed Rust in a long time and never used
`nom` before.

[1]: https://github.com/attwad/python-osc
[2]: https://github.com/D1plo1d/rosc/tree/feature/no_std

klingtnet committed:
Simplify address matching by providing a Matcher

A Matcher is constructed from an OSC address pattern and will use a
number of precompiled regular expression to match against a concrete OSC
address.  Regular expression match quite well to the pattern rules given
by the OSC specification and are used by other popular OSC libraries for
this purpose, e.g. [python-osc][1].

I chose `nom` as a parser combinator library to parse OSC addresses and
OSC address patterns reliably and efficiently.

@DrLuke please do not take this as a criticism of your implementation, I
am very about happy your contribution!  Actually, I would be happy if
_you_ could review this change so we can merge it back into your PR and
eventually release this feature.

My GitHub feed showed me that @D1plo1d is working on a [`no_std`
feature][2] which also adds `nom`.  I would be happy to receive a review
from @D1plo1d as well, hoping to reduce merge conflicts with their
feature.

Please be kind, I haven't programmed Rust in a long time and never used
`nom` before.

[1]: https://github.com/attwad/python-osc
[2]: https://github.com/D1plo1d/rosc/tree/feature/no_std

SQUASH: Add a couple more test cases

This revealed some errors, e.g. empty pattern parts are excepted like
`//foo/` which should be rejected.

klingtnet committed:
Support expect_err for Matcher by deriving Debug

DrLuke commented:
Hey, awesome PR, at first glance that looks like a way better approach!

I'll try to get a review done on sunday, I don't think I can get it in earlier.

klingtnet commented:
> Hey, awesome PR, at first glance that looks like a way better approach!
> 
> I'll try to get a review done on sunday, I don't think I can get it in earlier.

No hurry, I will be on vacation in the upcoming week.  I never used `nom` before so I hope that we will see some feedback from @D1plo1d who seems to have a bit of experience with the parser combinator library.

DrLuke commented on src/address.rs:
```diff
@@ -1,174 +1,132 @@
+    /// - `*` matches zero or more characters
+    /// - `[a-z]` are basically regex [character classes](https://www.regular-expressions.info/charclass.html)
+    /// - `{foo,bar}` is an alternative, matching either `foo` or `bar`
+    /// - everything else is matches literally
```
```suggestion
    /// - everything else is matched literally
```

DrLuke commented on src/address.rs:
````diff
@@ -1,174 +1,132 @@
+    /// ```
+    pub fn new(pattern: &str) -> Result<Self, OscError> {
+        let res = parse_address_pattern(pattern)?;
+        Ok(Matcher { res: res })
````
```suggestion
        Ok(Matcher { res })
```

DrLuke commented on src/address.rs:
```diff
@@ -1,174 +1,132 @@
-
-    let re = Regex::new(pattern.as_str()).unwrap();
+fn wrap_with(s: &str, pre: &str, post: &str) -> String {
+    pre.to_string() + s + post.into()
```
```suggestion
    pre.to_string() + s + post
```

DrLuke commented on src/address.rs:
```diff
@@ -1,174 +1,132 @@
-    }
+fn parse_address_pattern_part(input: &str) -> IResult<&str, &str> {
+    // TODO: the second parser, is_not, should require a non empty match
+    preceded(char('/'), is_not(" \n\t\r/"))(input)
```
```suggestion
    preceded(char('/'), take_till1(|c| " \n\t\r/".contains(c)))(input)
```

This will take characters until the condition is met, but at least 1

DrLuke commented on src/address.rs:
```diff
@@ -1,174 +1,132 @@
-    let method_addr_parts: Vec<&str> = method_addr.split('/').collect();
+fn parse_address_part(input: &str) -> IResult<&str, &str> {
+    // TODO: the second parser, is_not, should require a non empty match
+    preceded(char('/'), is_not(" \t\r\n#*,/?[]{}"))(input)
```
```suggestion
    preceded(char('/'), take_till1(|c| " \t\r\n#*,/?[]{}".contains(c)))(input)
```
This will take characters until the condition is met, but at least 1

DrLuke commented on tests/address_test.rs:
```diff
@@ -1,267 +1,47 @@
+    let expected_err = "bad OSC address pattern: bad address pattern";
+    assert_eq!(Matcher::new("").unwrap_err().to_string(), expected_err);
+    assert_eq!(Matcher::new("/").unwrap_err().to_string(), expected_err);
+    assert_eq!(Matcher::new("//empty/parts/").unwrap_err().to_string(), expected_err); // FIX: will fail
```
This test doesn't fail

```suggestion
    assert_eq!(Matcher::new("//empty/parts/").unwrap_err().to_string(), expected_err);
```

DrLuke commented on src/address.rs:
```diff
@@ -1,174 +1,132 @@
-
-    Ok(())
+use nom::branch::alt;
+use nom::bytes::complete::{is_not, tag};
```
```suggestion
use nom::bytes::complete::{is_not, tag, take_till1};
```

DrLuke reviewed (approved):
When thinking about performance I'm not sure if having a seperate Regex instance for each part is the way to go, but maybe this is an optimization for future PRs. Looks good to me otherwise, and nom seems like a pretty good choice of library!

Feel free to apply the suggestions I made if you want and let me know when I can merge.

klingtnet committed:
Apply review suggestions

s/matches/matched/

Co-authored-by: Lukas <Lukas@Jackowski.de>

Remove redundant field name

Co-authored-by: Lukas <Lukas@Jackowski.de>

Remove redundant .into()

Co-authored-by: Lukas <Lukas@Jackowski.de>

Use take_till1 to require a non-empty match

Co-authored-by: Lukas <Lukas@Jackowski.de>

Use take_till1 to require a non-empty match

Co-authored-by: Lukas <Lukas@Jackowski.de>

Use take_till1 to require a non-empty match

Co-authored-by: Lukas <Lukas@Jackowski.de>

Add missing import of take_till1

Co-authored-by: Lukas <Lukas@Jackowski.de>

klingtnet commented:
@DrLuke Thank you very much for the review and please excuse my late reply because I was on vacation and started a new job this week.  Besides, I applied all the suggestions.

> When thinking about performance I'm not sure if having a seperate Regex instance for each part is the way to go, but maybe this is an optimization for future PRs.

This is a valid point and, as you already said, something one can address in an upcoming PR.

klingtnet commented:
This PR is fine from my side and ready  to merge.  Feel free to do that whenever you're ready.
